heareval/score.py

```diff
@@ -0,0 +1,189 @@
+"""
+Common utils for scoring.
+"""
+from functools import partial
+from typing import Any, Callable, Dict, List, Optional
+
+import numpy as np
+import pandas as pd
+import sed_eval
+from dcase_util.containers import MetaDataContainer
+
+
+def label_vocab_as_dict(df: pd.DataFrame, key: str, value: str) -> Dict:
+    """
+    Returns a dictionary of the label vocabulary mapping the label column to
+    the idx column. key sets whether the label or idx is the key in the dict. The
+    other column will be the value.
+    """
+    if key == "label":
+        # Make sure the key is a string
+        df["label"] = df["label"].astype(str)
+        value = "idx"
+    else:
+        assert key == "idx", "key argument must be either 'label' or 'idx'"
+        value = "label"
+    return df.set_index(key).to_dict()[value]
+
+
+class ScoreFunction:
+    """
+    A simple abstract base class for score functions
+    """
+
+    # TODO: Remove task_metadata since we don't use it much
+    def __init__(
+        self,
+        task_metadata: Dict,
+        label_to_idx: Dict[str, int],
+        name: Optional[str] = None,
+    ):
+        self.task_metadata = task_metadata
+        self.label_to_idx = label_to_idx
+        if name:
+            self.name = name
+
+    def __call__(self, predictions: Any, targets: Any, **kwargs) -> float:
+        """
+        Compute the score based on the predictions and targets. Returns the score.
+        """
+        raise NotImplementedError("Inheriting classes must implement this function")
+
+    def __str__(self):
+        return self.name
```

Comment:
This looks like it will fail for the event detection methods below that don't have names defined. If name is optional and potentially not defined, do you want to return the name of the class if name isn't a defined attribute?

Reply:
dcase is currently way broken in this PR in a variety of ways, I am fixing it in upcoming PRs